iceoryx_examples/ara_com/ara/runtime.hpp

```diff
@@ -0,0 +1,171 @@
+                .or_else([](auto) {
+                    std::cerr << "unable to attach discovery" << std::endl;
+                    std::exit(EXIT_FAILURE);
+                });
+        }
+
+        return ara::com::FindServiceHandle({serviceIdentifier, instanceIdentifier});
+    }
+
+    void StopFindService(ara::com::FindServiceHandle handle) noexcept
+    {
+        auto iter = m_callbacks.begin();
+        for (; iter != m_callbacks.end(); iter++)
+        {
+            if (iter->second.GetServiceId() == handle.GetServiceId()
+                && iter->second.GetInstanceId() == handle.GetInstanceId())
+            {
+                break;
+            }
+        }
+        if (iter != m_callbacks.end())
+        {
+            m_callbacks.erase(iter);
+        }
+
+        if (m_callbacks.empty())
+        {
+            m_listener.detachEvent(m_discovery, iox::runtime::ServiceDiscoveryEvent::SERVICE_REGISTRY_CHANGED);
+        }
+    }
+
+  private:
+    explicit Runtime() noexcept = default;
+
+    bool verifyThatServiceIsComplete(ara::com::ServiceHandleContainer<com::ProxyHandleType>& container)
+    {
+        // The service level of AUTOSAR Adaptive is not available in iceoryx, instead every publisher and server is
+        // considered as a service. A ara::com binding implementer would typically query the AUTOSAR meta model here, to
+        // find out if all event, fields and methods of a service are available. For the example we assume that the
+        // 'MinimalSkeleton' service is complete when the container contains the three iceoryx services:
+        //
+        // 1. EventPublisher: MinimalSkeleton, Instance, Event
+        // 2. FieldPublisher: MinimalSkeleton, Instance, Field
+        // 3. MethodServer:   MinimalSkeleton, Instance, Method
+
+        if (container.size() == 3U)
+        {
+            return true;
+        }
+        return false;
+    }
+
+    static void invokeCallback(iox::runtime::ServiceDiscovery*, Runtime* self)
+    {
+        // 1) Has the availability of one of the registered services changed?
+        // 2) If yes, call the user-defined callback
+        for (auto& callback : self->m_callbacks)
+        {
+            auto container =
+                self->FindService(callback.second.m_serviceIdentifier, callback.second.m_instanceIdentifier);
+            // if (container.empty())
+            // {
+            //     continue;
+            // }
+            (callback.first)(container,
+                             ara::com::FindServiceHandle(
+                                 {callback.second.m_serviceIdentifier, callback.second.m_instanceIdentifier}));
+        }
+    }
+
+    iox::runtime::ServiceDiscovery m_discovery;
+    iox::popo::Listener m_listener;
+    iox::cxx::vector<iox::cxx::pair<ara::com::FindServiceHandler<com::ProxyHandleType>, ara::com::FindServiceHandle>,
```

Comment:
A vector is clearly the wrong data structure for this (we remove anywhere). We need at least a list, but actually something construction on top of vector where we manage indices (=the free handles) and an array where we can set the callback at a specific index is ideal.

Reply:
Sure, `vector` is not optimal. But remember this is an example ;- ) IMHO that's for an example sufficient.